tests/test_appipe.py

```diff
@@ -21,22 +21,142 @@
+            mockDataRef = Mock(dafPersist.ButlerDataRef)
+            mockDataRef.dataId = dict(dataId, **rest)
+            return mockDataRef
+        butlerPatcher = patch.object(self.butler, "dataRef",
+                                     side_effect=makeMockDataRef)
+        butlerPatcher.start()
+        self.addCleanup(butlerPatcher.stop)
+        self.dataId = {"visit": 413635, "ccdnum": 42}
+
+    @contextlib.contextmanager
+    def mockPatchSubtasks(self, task):
+        """Make mocks for all the ap_pipe subtasks.
+
+        Parameters
+        ----------
+        task : `lsst.ap.pipe.ApPipeTask`
+            the task whose subtasks will be mocked
+
+        Yields
+        ------
+        subtasks : `lsst.pipe.base.Struct`
+            All mocks created by this context manager, including:
+
+            ``ccdProcessor``
+            ``differencer``
+            ``dpddifier``
+            ``associator``
+            ``forcedSource``
+                a mock for the corresponding subtask. Mocks do not return any
+                particular value, but have mocked methods that can be queried
+                for calls by ApPipeTask
+        """
+        with patch.object(task, "ccdProcessor") as mockCcdProcessor:
+            with patch.object(task, "differencer") as mockDifferencer:
+                with patch.object(task, "diaSourceDpddifier") as mockDpddifier:
+                    with patch.object(task, "associator") as mockAssociator:
+                        with patch.object(task, "diaForcedSource") as mockForcedSource:
+                            yield pipeBase.Struct(ccdProcessor=mockCcdProcessor,
+                                                  differencer=mockDifferencer,
+                                                  dpddifier=mockDpddifier,
+                                                  associator=mockAssociator,
+                                                  forcedSource=mockForcedSource)
+
+    def testGenericRun(self):
+        # test the normal workflow of each ap_pipe step
+        task = ApPipeTask(self.butler, config=self.config)
+        inputRef = self.butler.dataRef("raw", **self.dataId)
+        with self.mockPatchSubtasks(task) as subtasks:
+            task.runDataRef(inputRef)
+            subtasks.ccdProcessor.runDataRef.assert_called_once()
+            subtasks.differencer.runDataRef.assert_called_once()
+            subtasks.associator.run.assert_called_once()
+            subtasks.forcedSource.run.assert_called_once()
+
+    def testReuseExistingOutput(self):
+        # test reuse keyword to ApPipeTask.runDataRef
+        task = ApPipeTask(self.butler, config=self.config)
+        inputRef = self.butler.dataRef("raw", **self.dataId)
+
+        # Mock guarantees that all "has this been made" tests pass,
+        # so skippable subtasks should actually be skipped
+        for skippable in [
+            ['ccdProcessor'],
+            ['ccdProcessor', 'differencer'],
+        ]:
```

Comment:
Loops in tests make the tests difficult to untangle. You're only doing two things here, so I think it would be clearer if you made the internals of this into a separate method (e.g. `checkSkippedTask`) and then call it:

```
self.checkSkippedTask(['ccdProcessor'])
self.checkSkippedTask(['ccdProcessor', 'differencer'])
```

Similarly for the loop below, although I'm not sure how to unravel that one.

Reply:
We fixed the first loop as suggested and added a message that identifies the subtask being tested for the second loop.